src/Query.jl

```diff
@@ -0,0 +1,195 @@
+  OrWhere(a::OrWhere, b) = OrWhere(a.conds, b)
+  OrWhere(a, b::OrWhere) = OrWhere(a, b.conds)
+  OrWhere(a, b) = new([a; b])
+end
+
+function Base.:|(a::S, b::T) where {T<:AbstractCondition, S<:AbstractCondition}
+  OrWhere(a, b)
+end
+
+mutable struct SQLACSetNode
+  from::Symbol
+  cond::Union{Vector{<:AbstractCondition}, Nothing}
+  select
+  SQLACSetNode(from::Symbol; cond=nothing, select=nothing) = new(from, cond, select)
+end
+
+function (w::WhereCondition)(node::SQLACSetNode)
+  push!(node.cond, AndWhere([w]))
+  node
+end
+
+function (ac::AbstractCondition)(node::SQLACSetNode)
+  push!(node.cond, ac)
+  node
+end
+
+function Base.:&(n::SQLACSetNode, a::AbstractCondition)
+  n.cond = n.cond & a
+  n
+end
+
+function Base.:|(n::SQLACSetNode, a::AbstractCondition)
+  n.cond = n.cond | a
+  n
+end
+
+From(table::Symbol) = SQLACSetNode(table; cond=AbstractCondition[], select=[])
+
+function From(tablecol::Pair{Symbol, Symbol})
+  SQLACSetNode(tablecol.first; cond=AbstractCondition[], select=[tablecol.second])
+end
+
+Where(lhs, op::Function, rhs) = WhereCondition(lhs, op, rhs)
+Where(lhs::Symbol, rhs::Function) = Where(lhs, |>, rhs)
+Where(lhs::Symbol, rhs) = Where(lhs, ∈, rhs)
+Where(lhs, rhs::Function) = Where(lhs, |>, rhs)
+
+function Select(sql::SQLACSetNode, columns::Vector)
+  push!(sql.select, columns...)
+  sql
+end
+
+function Select(columns...)
+  sql -> Select(sql, [columns...])
+end
+
+function process_wheres(conds::Vector{<:AbstractCondition}, acset)
+  isempty(conds) && return [Colon()]
+  process_where.(conds, Ref(acset))
+end
+
+function process_where(cond::WhereCondition, acset::ACSet)
+  values = get(acset, cond.lhs)
+  map(values) do value
+    @match (value, cond.rhs) begin
+      (_, ::SQLACSetNode)            => cond.op(value, cond.rhs(acset)[1].second)
+      (::Tuple,          ::Function) => cond.rhs(value...)
+      (::AbstractVector, ::Function) => cond.rhs(value...)
+      (_, ::Function)                => cond.rhs(value)
+      (_, ::Vector)                  => cond.op(iterable(value)..., cond.rhs)
+      _                              => cond.op(iterable(value)..., [cond.rhs])
+    end
+  end
+end
+
+function process_where(w::OrWhere, acset::ACSet)
+  isempty(w.conds) && return nothing
+  reduce((x,y) -> x .| y, process_wheres(w.conds, acset))
+end
+
+function process_where(w::AndWhere, acset::ACSet)
+  isempty(w.conds) && return nothing
+  reduce((x,y) -> x .& y, process_wheres(w.conds, acset))
+end
+
+function process_select(q::SQLACSetNode, acset::ACSet, result::AbstractVector)
+  isempty(q.select) && return q.from => result
+  map(q.select) do select
+    to_name(select) => @match select begin
+      ::Val{T} where T           => [T for _ in eachindex(result)]
+      ::Symbol                   => get(acset, select, result)
+      ::Pair{Symbol, <:Function} => get(acset, select.first, result) .|> select.second
+    end
+  end
+end
+
+abstract type AbstractQueryFormatter end
```

Comment:
Need some documentation about what interface (informally, no GATs needed!) implementations of this abstract type should satisfy. Also very short descriptions of the structs below as docstrings.